src/core/qgspointlocator.cpp

```diff
@@ -336,57 +336,29 @@ static QgsPointLocator::MatchList _geometrySegmentsInRect( QgsGeometry *geom, co
   QgsGeometry straightGeom = geom->convertToType( QgsWkbTypes::LineGeometry, true );
   straightGeom.convertToStraightSegment();
 
-  QgsWkbTypes::Type wkbType = straightGeom.wkbType();
-
   if ( ( straightGeom.type() != QgsWkbTypes::LineGeometry ) && ( !straightGeom.isMultipart() ) )
     return lst;
 
-  QByteArray wkb( straightGeom.asWkb() );
-  if ( wkb.isEmpty() )
-    return lst;
-
   _CohenSutherland cs( rect );
 
-  QgsConstWkbPtr wkbPtr( wkb );
-  wkbPtr.readHeader();
-
-
-  bool hasZValue = QgsWkbTypes::hasZ( wkbType );
-  bool hasMValue = QgsWkbTypes::hasM( wkbType );
-
-  int nLines;
-  wkbPtr >> nLines;
-  for ( int linenr = 0, pointIndex = 0; linenr < nLines; ++linenr )
+  int pointIndex = 0;
+  QgsAbstractGeometry::vertex_iterator it = straightGeom.vertices_begin();
+  QgsPointXY prevPoint( *it );
```

Comment:
Is this safe to do on an empty linestring (or is this caught somewhere)?

Reply:
Yes, it's excluded before, at `if ( ( straightGeom.type() != QgsWkbTypes::LineGeometry ) && ( !straightGeom.isMultipart() ) )`, since it's not a LineGeometry and not a multipart geom.

Reply:
Are you sure?

```
>>> QgsGeometry(QgsLineString()).type() == QgsWkbTypes.LineGeometry
True
```

Reply:
Hmmm, tested with a Wkt Empty LineStringZ (postgis for a real case)...

```
>>> g = QgsGeometry.fromWkt("LineStringZ EMPTY")
>>> g
<QgsGeometry: null>
>>> g.type() != QgsWkbTypes.LineGeometry
True
```
It reminds me of https://issues.qgis.org/issues/20753 (and others)

As a precaution, I'll return an empty match if the geometry is empty.

Reply:
That sounds like an issue on the wkt parser side. Or do you think there is a good reason to drop type information if there are no vertices around (and analogue, should a single-vertex line be of type point and would adding an additional vertex make this a multipoint?)

Reply:
Yes, there is some issues with wkt and maybe with some case for our geometries.
An empty linestring is a linestring type. Example with postgis:
`SELECT ST_GeometryType(ST_GeomFromText('LINESTRING EMPTY'))` returns `ST_LineString`. It's an empty set for LineString type according to the standard [1].

AFAIK a LineString with only point, should not be possible. 

A curve must have a start and an end point[2], so at least two points (which can be the same). 

But, the BNF notation for a LineString[1] is:
`<linestring text> ::= <empty set> | <left paren> <point> {<comma> <point>}* <right paren>` so an empty set or a line with a point and (*) 0 or more point... 
Maybe is an error, since for 
`<linestring m text> ::= <empty set> | <left paren> <point m> {{<comma> <point m>}+ <right paren>` [2] so a line with a point m and (+) 1 or more point.

I have to look at the standard in more detail, but for example postgis does not allow this:
`SELECT ST_GeomFromText('LineString(0 1)')` <- "geometry requires more points"

But QGIS yes (with python):
````
>>> line = QgsLineString([QgsPoint(0,1)])
>>> line
<QgsLineString: LineString (0 1)>
>>> line.startPoint()
<QgsPoint: Point (0 1)>
>>> line.endPoint()
<QgsPoint: Point (0 1)>
```

However, I think that this is beyond the scope of this PR. I suggest that we work on these topics in separate PR and issues.

But, again as a precaution, I will add a check on the number of points in the resulting line.

[1] http://portal.opengeospatial.org/files/?artifact_id=25355 p55
[2] http://portal.opengeospatial.org/files/?artifact_id=25355 p22
[3] http://portal.opengeospatial.org/files/?artifact_id=25355 p58

Reply:
> But, again as a precaution, I will add a check on the number of points in the resulting line.

That very much sounds like the correct solution here :+1: 

---------

All other considerations boil down to *when is the right time to detect invalid geometries*. This involves data providers and programatically created geometries and definitely is out of scope for this PR.